plugins/registry.js

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import reducerRegistry from 'mattermost-redux/store/reducer_registry';
+
+import store from 'stores/redux_store.jsx';
+import {ActionTypes} from 'utils/constants.jsx';
+import {generateId} from 'utils/utils.jsx';
+
+function dispatchPluginComponentAction(name, pluginId, component, id = generateId()) {
+    store.dispatch({
+        type: ActionTypes.RECEIVED_PLUGIN_COMPONENT,
+        name,
+        data: {
+            id,
+            pluginId,
+            component,
+        },
+    });
+
+    return id;
+}
+
+export default class PluginRegistry {
+    constructor(id) {
+        this.id = id;
+    }
+
+    // Register a component at the root of the channel view of the app.
+    // Accepts a React component. Returns a unique identifier.
+    registerRootComponent = (component) => {
+        return dispatchPluginComponentAction('Root', this.id, component);
+    }
+
+    // Register a component in the first section of the profile popover (hovercard), below the user's name.
+    // Accepts a React component. Returns a unique identifier.
+    registerPopoverSection1Component = (component) => {
+        return dispatchPluginComponentAction('PopoverSection1', this.id, component);
+    }
+
+    // Register a component in the second section of the profile popover (hovercard), below the user's email.
+    // Accepts a React component. Returns a unique identifier.
+    registerPopoverSection2Component = (component) => {
+        return dispatchPluginComponentAction('PopoverSection2', this.id, component);
+    }
+
+    // Register a component in the third section of the profile popover (hovercard), below the send message button.
+    // Accepts a React component. Returns a unique identifier.
+    registerPopoverSection3Component = (component) => {
+        return dispatchPluginComponentAction('PopoverSection3', this.id, component);
+    }
+
+    // Register a custom button component to add to the channel header. If there are more than one
```

Comment:
With the other API, we do some preamble around getting the button set up and applying our styling to it. This API is meant to let you ignore all that and style the button exactly how you'd like

Reply:
Ah, I definitely didn't parse that well. This type of API seems potentially problematic, since it's easy to test having just passed one parameter and skip the other not realizing it will break in a mobile context.

I think I had imagined a more structured input to the channel headers (and main menu) as described above, which automatically generated dropdowns or buttons based on the plugin's actions and icons.